cts/scheduler/xml/promoted-probed-score.xml

```diff
@@ -623,7 +623,7 @@
       </lrm>
       <transient_attributes id="hypatia-corosync.nevis.columbia.edu">
         <instance_attributes id="status-hypatia-corosync.nevis.columbia.edu">
-          <nvpair id="status-hypatia-corosync.nevis.columbia.edu-promoted-AdminDrbd.0" name="master-AdminDrbd:0" value="5"/>
+          <nvpair id="status-hypatia-corosync.nevis.columbia.edu-promoted-AdminDrbd" name="master-AdminDrbd" value="5"/>
```

Comment:
Even when I make it globally unique, the promotion scores that are instance-specific don't influence assignment. So Pacemaker assigns the instances however it wants to, and then if the "wrong" instance winds up on the "wrong" node, these promotion scores have no effect. If we want the instance-specific promotion scores to influence assignment, that's gonna be a separate task :)

I lean toward making "strip the instance number if it's an anonymous clone" a separate task too, if we want to do that. Testing promotion scores wasn't the purpose of this test. The instance-specific promotion scores just happened to break the test after a scheduler code change caused the instance assignments to flip.

---

It's not entirely clear what we'd want to match. We don't read through the list of meta attributes while fetching promotion scores. Rather, we look up the specific one we want:
```
char *
pcmk_promotion_score_name(const char *rsc_id)
{
    ...
    return crm_strdup_printf("master-%s", rsc_id);
}

static const char *
promotion_attr_value(const pe_resource_t *rsc, const pe_node_t *node,
                     const char *name)
{
    ,,,
    attr_name = pcmk_promotion_score_name(name);
    attr_value = pe_node_attribute_calculated(node, attr_name, rsc, node_type);
```

Even if we wanted to, we couldn't efficiently look up a regex `attr_name` instead of a raw string `attr_name`. If a colon is an allowed character in a resource ID, then that might not be robust to corner cases anyway.

I suppose we could iterate over the whole `rsc->meta` hash table and do a regex comparison on each value...

Or maybe there's something simple that hasn't occurred to me. I'm pretty sleepy, just addressing review before bed

Reply:
Could special-case the promotion score when unpacking resource meta attributes from XML in the first place, I suppose

Reply:
> I don't think promotion scores ever influence assignment

In a sense they do. `pcmk__add_promotion_scores()` gets called at the beginning of `pcmk__clone_assign()`, and it adds the promotion score to the allowed node's score and priority.

You could make the case that if there's an instance-specific promotion score like we had in this test -- particularly if the clone is unique -- it should influence which instance gets assigned where. I haven't looked into how easy that would be.

This behavior had a tangible negative effect before I dropped the `:<num>` suffix from the attributes. Each instance had a promotion score on only one node, so if they were assigned to the opposite nodes, neither instance was allowed to promote. In turn, every other resource in the cluster has a mandatory colocation with the promoted role of `AdminClone`. So when assignment flipped to the "wrong" nodes, the clone instances couldn't promote, and nothing else could start.

That's arguably still undesirable even for unique clones.

> Promotion scores are node-specific, so there's no wrong node :)

See above, I'm using "wrong" loosely haha

> Right, it would just be a good side test. Sometimes we throw in trivial stuff that isn't worth its own test

Agreed

> rsc_id (and OCF_RESOURCE_INSTANCE) for globally unique instances includes the instance number (the function is called for instances)

Right. We're looking up the attribute by exact name though. Suppose we have the following meta attributes set (values unimportant):
* `master-AdminDrbd`
* `master-AdminDrbd:0`
* `master-AdminDrbd:1`
* `master-AdminDrbd:9`
* `master-AdminDrbd:999`

For a unique clone instance 0, we'd have to match against either `master-AdminDrbd` or `master-AdminDrbd:0` (letting either the most specific one or the last one win). For anonymous clones, we'd have to match against every single one of them, letting the last one win. There's no way to do a `g_hash_table_lookup()` by regex.

Maybe we could strip the instance number at unpack time for anonymous clones, idk. Not sure offhand if anything would need to change for unique clones.

Reply:
> > I don't think promotion scores ever influence assignment
> 
> In a sense they do. `pcmk__add_promotion_scores()` gets called at the beginning of `pcmk__clone_assign()`, and it adds the promotion score to the allowed node's score and priority.

Ah, I forgot about that

> You could make the case that if there's an instance-specific promotion score like we had in this test -- particularly if the clone is unique -- it should influence which instance gets assigned where. I haven't looked into how easy that would be.

I would expect that currently only instance-specific promotion scores that would be considered for unique clones, and only non-specific promotion scores would be considered for anonymous clones
 
> This behavior had a tangible negative effect before I dropped the `:<num>` suffix from the attributes. Each instance had a promotion score on only one node, so if they were assigned to the opposite nodes, neither instance was allowed to promote. In turn, every other resource in the cluster has a mandatory colocation with the promoted role of `AdminClone`. So when assignment flipped to the "wrong" nodes, the clone instances couldn't promote, and nothing else could start.
> 
> That's arguably still undesirable even for unique clones.

Nope, that makes sense. Promotion scores are most commonly set by resource agent monitor calls (probes, initially). Each instance of a unique clone is probed separately on each node. I can't remember for sure now, but I believe the agent gets the resource name (OCF_RESOURCE_INSTANCE) with an instance number for unique clones and without an instance number for anonymous clones, so promotion scores should be set as expected.
 
> > Promotion scores are node-specific, so there's no wrong node :)
> 
> See above, I'm using "wrong" loosely haha
> 
> > Right, it would just be a good side test. Sometimes we throw in trivial stuff that isn't worth its own test
> 
> Agreed
> 
> > rsc_id (and OCF_RESOURCE_INSTANCE) for globally unique instances includes the instance number (the function is called for instances)
> 
> Right. We're looking up the attribute by exact name though. Suppose we have the following meta attributes set (values unimportant):
> 
>     * `master-AdminDrbd`
> 
>     * `master-AdminDrbd:0`
> 
>     * `master-AdminDrbd:1`
> 
>     * `master-AdminDrbd:9`
> 
>     * `master-AdminDrbd:999`
> 
> 
> For a unique clone instance 0, we'd have to match against either `master-AdminDrbd` or `master-AdminDrbd:0` (letting either the most specific one or the last one win). For anonymous clones, we'd have to match against every single one of them, letting the last one win. There's no way to do a `g_hash_table_lookup()` by regex.
> 
> Maybe we could strip the instance number at unpack time for anonymous clones, idk. Not sure offhand if anything would need to change for unique clones.

It would be reasonable to check only the numbered attributes for unique clones and only the non-numbered attribute for anonymous clones, and I hope that's what the current code does. It would be nice but not essential to fall back to the unnumbered attribute for unique clones and strip instance numbers for anonymous clones, mainly to handle the case where globally-unique was toggled in the configuration.